### ISSUE TITLE
Allow to print PDF

### ISSUE THREAD

**Issue by VadimDez** [VadimDez/ng2-pdf-viewer#69]

**Comment by skormel** [comment]
When is going to be implementedthis feature? 

Best Regards.

**Comment by fad250** [comment]
Is this going to be implemented?

**Comment by sivahanuman** [comment]
hi, can we print the pdf document ? documentation please.

**Comment by hesampour** [comment]
Hi, this is a very important feature.
would you please add this feature??

**Comment by rohanrp** [comment]
I got it to print using 

const blobUrl  = window.URL.createObjectURL((this.srcBinary);
const iframe = document.createElement('iframe');
iframe.style.display = 'none';
iframe.src = blobUrl;
document.body.appendChild(iframe);
iframe.contentWindow.print();

**Comment by ferrarinim** [comment]
> I got it to print using
> 
> const blobUrl = window.URL.createObjectURL((this.srcBinary);
> const iframe = document.createElement('iframe');
> iframe.style.display = 'none';
> iframe.src = blobUrl;
> document.body.appendChild(iframe);
> iframe.contentWindow.print();

I tryed your code but it doesn't print anything...
Did you solve?

**Comment by dhamodharaanbm** [comment]
@VadimDez  I need to print the document with the data on top of it ,  would you please add this feature??

**Comment by seopei** [comment]
I am using Print.js (https://printjs.crabbly.com/) in order to print to document: 
```
    let binary = '';
    const len = this.documentSrc.byteLength;
    for (let i = 0; i < len; i++) {
      binary += String.fromCharCode(this.documentSrc[i]);
    }
    printJS({
      printable: btoa(binary),
      type: 'pdf',
      base64: true
    });
```

**Comment by juliengit2** [comment]
Is this going to be implemented?

**Comment by kiranhpe** [comment]
> I got it to print using
> 
> const blobUrl = window.URL.createObjectURL((this.srcBinary);
> const iframe = document.createElement('iframe');
> iframe.style.display = 'none';
> iframe.src = blobUrl;
> document.body.appendChild(iframe);
> iframe.contentWindow.print();

it works for me without any issue

**Comment by bhawin-boom** [comment]
> I got it to print using
> 
> const blobUrl = window.URL.createObjectURL((this.srcBinary);
> const iframe = document.createElement('iframe');
> iframe.style.display = 'none';
> iframe.src = blobUrl;
> document.body.appendChild(iframe);
> iframe.contentWindow.print();

This only works in chrome  , failed in Firefox , edge and IE

**Comment by bhawin-boom** [comment]
@VadimDez  , are we implementing this one

**Comment by jpike88** [comment]
I think you should close this. There's no way to reliably trigger a print from the plugin itself.

Google drive even works around this by simply opening the document directly in a new tab, and letting the user decide what to do. It doesn't even trigger the print dialog.

**Comment by rohanrp** [comment]
PDF.js allows you to print without opening new tab
see https://mozilla.github.io/pdf.js/web/viewer.html

**Comment by jpike88** [comment]
That's probably using a trick with iframes or something, and the outcome is less than ideal because of all the footer and extra stuff the browser throws in there. If you're trying to print a PDF, you probably don't want that stuff present...

**Comment by rohanrp** [comment]
It prints it perfectly. Here is a screenshot from Ctrl + P. My above code does the same thing except it doesn't work in IE11

<img width="514" alt="Capture" src="https://user-images.githubusercontent.com/14039214/89010733-7f9d0f80-d352-11ea-8ae7-c0c7eb2abef3.PNG">

**Comment by jpike88** [comment]
This is in Safari

<img width="473" alt="Screen Shot 2020-07-31 at 3 24 57 PM" src="https://user-images.githubusercontent.com/9585787/89010873-ff6eae00-d341-11ea-8370-fd27ee95f7bc.png">

**Comment by rohanrp** [comment]
that's because safari treats the document as a webpage instead of PDF. If it helps, the user can disable print headers and footers in Safari browser.

https://discussions.apple.com/thread/3046001

**Comment by jpike88** [comment]
Well that's a non-starter then, because users aren't that well versed in their browser configuration to know how to do that. It creates pain for support. 

Their browser is already capable of viewing and printing PDFs cleanly (even IE11, which uses what reader is available, typically Acrobat), which is why something like Google drive, even with a document previewer, chooses to open in a new tab when print is clicked.

**Comment by rohanrp** [comment]
Well fair enough.

One of the reasons many people use ng2-pdf-viewer or pdf.js is because companies want to restrict sensitive documents from being printed or stored on the user's computer (although you can hack around and download the PDF via direct url). 

So the print option is not a very strong use case for this library.

**Comment by stephanrauh** [comment]
I might provide a pull request. Printing isn't that difficult - apart from the prob

**Comment by stephanrauh** [comment]
@VadimDez I've got a working prototype. It's not perfect - one of my test files is printed on 19 instead of 18 pages, but it's already good enough for most use cases.

I'm only worried about copyright.  My implementation is based on Mozilla's print algorithm. I've ported it to TypeScript and simplified it a lot, so according to German law, I'm the copyright owner. But I'm not sure if that's also true by international law. So it might be difficult to publish it under an MIT license.
